### PR TITLE
[react-cache] Add typings for react-cache

### PR DESCRIPTION
**Pull request by skovy**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/react-cache/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+// Type definitions for react-cache 2.0
+// Project: https://github.com/facebook/react/tree/master/packages/react-cache
+// Definitions by: Spencer Miskoviak <https://github.com/skovy>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export interface Thenable<Value> {
+  then(resolve: (value: Value) => any, reject: (reason: any) => any): any;
+}
```
Comment (Kovensky) [inline types/react-cache/index.d.ts]: In TS there's already a global declaration for Thenable which is more type-safe than this (it's `PromiseLike<T>`)

File: types/react-cache/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+export interface Resource<Input, Value> {
+  read(key: Input): Value;
+  preload(key: Input): void;
+}
+
+export function unstable_createResource<Input, Value>(
+  fetch: (input: Input) => Thenable<Value>,
+  maybeHashInput?: (input: Input) => string | number
```
Comment (Kovensky) [inline types/react-cache/index.d.ts]: Interesting that in the Flow types they have this as a generic input, but don't actually use its genericness.

In TS, that being generic would have the effect of inferring literal types for the return, but that doesn't seem productive for a hash function either.

**Comment by skovy** [comment]
Thanks for the feedback @Kovensky! I've updated it to use `PromiseLike`. Are there any other changes around `maybeHashInput` you would suggest making?

**Comment by Kovensky** [comment]
I was more trying to understand why the Flow types had a for the key result, it's there but it's not used.

It would be good to add a comment that `unstable_createResource`'s result cannot be destroyed / would leak memory if misused.

**Review by Jessidhia — APPROVED** [review]

**Comment by PranavSenthilnathan** [comment]
Congratulations on your first DefinitelyTyped contribution! Thank you for being a part of the community!

**Pull request merged**